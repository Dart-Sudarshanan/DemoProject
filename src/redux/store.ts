import { configureStore } from "@reduxjs/toolkit";
import authSlice from "./slices/authSlice";

import { reducer as formReducer } from "redux-form";
import userSlice from "./slices/userSlice";
import counterSlice from "./slices/counterSlice";

const store = configureStore({
  reducer: {
    form: formReducer,
    userAuth: authSlice.reducer,
    userList: userSlice,
    counterList: counterSlice,
  },
});

export type ApplicationState = ReturnType<typeof store.getState>;

export const authUser = (state: ApplicationState) => state.userAuth;
export const userList = (state: ApplicationState) => state.userList;
export const counterList = (state: ApplicationState) => state.counterList;

export { store };
