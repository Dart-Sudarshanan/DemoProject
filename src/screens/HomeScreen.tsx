import React, { useEffect } from "react";

import { StyleSheet, Text, View, FlatList, TouchableOpacity } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import { fetchUser } from "../redux/slices/userSlice";
import { userList } from "../redux/store";

const UserBlock = ({ data }: { data: any }, ...props: any): React.ReactElement => {
  return (
    <View style={styles.list} key={"user-list-" + data.id}>
      <View style={styles.listRow} key={"row_1_" + data.id}>
        <View style={styles.column}>
          <Text style={styles.listHeading}>Name:</Text>
          <Text>{data.name}</Text>
        </View>
        <View style={styles.column}>
          <Text style={styles.listHeading}>Gender:</Text>
          <Text ellipsizeMode={"tail"}>{data.gender}</Text>
        </View>
      </View>
      <View style={styles.listRow} key={"row_2_" + data.id}>
        <View style={{ padding: 10 }}>
          <Text style={styles.listHeading}>Email:</Text>
          <Text>{data.email}</Text>
        </View>
      </View>
    </View>
  );
};

const HomeScreen = (): React.ReactElement => {
  const dispatch = useDispatch();
  const { users, isLoading, hasError } = useSelector(userList);

  useEffect(() => {
    dispatch(fetchUser(users.page));
  }, [dispatch]);

  const getUserList = () => {
    dispatch(fetchUser(users.page + 1));
  };

  const renderUsers = () => {
    if (hasError) {
      return (
        <View>
          <Text>Some error has occured</Text>
        </View>
      );
    } else if (users) {
      if (users.data.length <= 0 && !isLoading) {
        return (
          <View>
            <Text>No data found.</Text>
          </View>
        );
      } else if (users.data.length > 0) {
        return (
          <FlatList
            key={"list-User"}
            data={users.data}
            onEndReached={getUserList}
            showsVerticalScrollIndicator={false}
            renderItem={({ item, index }) => (
              <>
                <TouchableOpacity key={item.id}>
                  <UserBlock key={"users-" + index} data={item} />
                </TouchableOpacity>
              </>
            )}
          />
        );
      }

      if (isLoading && users.page === 1) {
        return (
          <View>
            <Text>Getting User list ...</Text>
          </View>
        );
      }
    }
  };

  return (
    <View style={styles.sectionContainer}>
      <View style={styles.contentWrapper}>
        <Text style={styles.text}>Users</Text>
        {renderUsers()}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  sectionContainer: {
    marginTop: 10,
    paddingHorizontal: 15,
  },
  contentWrapper: {
    justifyContent: "center",
  },
  text: {
    fontSize: 24,
    fontWeight: "400",
  },
  button: {
    paddingHorizontal: 40,
    paddingVertical: 10,
    borderRadius: 4,
    backgroundColor: "#4c51e6",
  },
  buttonTxt: {
    fontSize: 18,
    fontWeight: "500",
    color: "#fff",
  },
  list: {
    backgroundColor: "#fff",
    marginVertical: 10,
    padding: 10,
    borderRadius: 5,
  },
  listRow: {
    flexDirection: "row",
  },
  column: {
    flexDirection: "column",
    padding: 10,
    width: "50%",
  },
  listHeading: {
    fontWeight: "700",
  },
});

export default HomeScreen;
