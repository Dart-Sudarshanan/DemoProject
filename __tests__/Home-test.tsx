/**
 * @format
 */

import "react-native";
import React from "react";

// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
import HomeScreen from "../src/screens/HomeScreen";
import { Provider } from "react-redux";
import { store } from "../src/redux/store";

describe("Home screen", () => {
  it("should returns no data found", async () => {
    jest.useFakeTimers();
    const tree = renderer.create(
      <Provider store={store}>
        <HomeScreen />
      </Provider>
    );
    expect(tree).toMatchSnapshot();
  });
});
