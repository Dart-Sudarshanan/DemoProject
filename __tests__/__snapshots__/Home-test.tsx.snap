// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen returns no data found 1`] = `
<View
  style={
    Object {
      "marginTop": 10,
      "paddingHorizontal": 15,
    }
  }
>
  <View
    style={
      Object {
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "400",
        }
      }
    >
      Users
    </Text>
    <View>
      <Text>
        No data found.
      </Text>
    </View>
  </View>
</View>
`;
